rnd/autogpt_builder/src/hooks/useAgentGraph.ts

```diff
@@ -0,0 +1,674 @@
+      return;
+    }
+    setUpdateQueue((prev) => {
+      prev.forEach((data) => {
+        // Skip already processed updates by checking
+        // if the data is in the processedUpdates array by reference
+        // This is not to process twice in react dev mode
+        // because it'll add double the beads
+        if (processedUpdates.current.includes(data)) {
+          return;
+        }
+        updateNodesWithExecutionData(data);
+        processedUpdates.current.push(data);
+        processedUpdates.current = processedUpdates.current.slice(-1000);
+      });
+      return [];
+    });
+  }, [updateQueue, nodesSyncedWithSavedAgent]);
+
+  // Handle user requests
+  useEffect(() => {
+    // Ignore none request
+    if (saveRunRequest.request === "none") {
+      return;
+    }
+    // Display error message
+    if (saveRunRequest.state === "error") {
+      if (saveRunRequest.request === "save") {
+        console.error("Error saving agent");
+      } else if (saveRunRequest.request === "run") {
+        console.error(`Error saving&running agent`);
+      }
+      // Reset request
+      setSaveRunRequest((prev) => ({
+        ...prev,
+        request: "none",
+        state: "none",
+      }));
+      return;
+    }
+    // When saving request is done
+    if (
+      saveRunRequest.state === "saving" &&
+      savedAgent &&
+      nodesSyncedWithSavedAgent
+    ) {
+      // Reset request if only save was requested
+      if (saveRunRequest.request === "save") {
+        setSaveRunRequest((prev) => ({
+          ...prev,
+          request: "none",
+          state: "none",
+        }));
+        // If run was requested, run the agent
+      } else if (saveRunRequest.request === "run") {
+        if (!validateNodes()) {
+          console.error("Validation failed; aborting run");
+          setSaveRunRequest({
+            request: "none",
+            state: "none",
+          });
+          return;
+        }
+        api.subscribeToExecution(savedAgent.id);
+        api.executeGraph(savedAgent.id);
+
+        setSaveRunRequest((prev) => ({
+          ...prev,
+          request: "run",
+          state: "running",
+        }));
+      }
+    }
+  }, [saveRunRequest, savedAgent, nodesSyncedWithSavedAgent]);
+
+  // Check if node ids are synced with saved agent
+  useEffect(() => {
+    // Check if all node ids are synced with saved agent (frontend and backend)
+    if (!savedAgent || nodes?.length === 0) {
+      setNodesSyncedWithSavedAgent(false);
+      return;
+    }
+    // Find at least one node that has backend id existing on any saved agent node
+    // This will works as long as ALL ids are replaced each time the graph is run
+    const oneNodeSynced = savedAgent.nodes.some(
+      (backendNode) => backendNode.id === nodes[0].data.backend_id,
+    );
+    setNodesSyncedWithSavedAgent(oneNodeSynced);
+  }, [savedAgent, nodes]);
+
+  const validateNodes = useCallback((): boolean => {
```

Comment:
I guess this is mostly unchanged from the previous implementation ?

Reply:
yes